### PR TITLE
feat: make reduce_u32 + split_u32 circuit friendly

### PR DESCRIPTION
**Pull request by jtguibas**
Makes reduce_u32 + split_u32 use a circuit-friendly decomposition. The previous implementation complicates the arithmetization in R1CS due to a decomposition based on the order of the field instead of just using bits.

### CODE REVIEW

File: field/src/helpers.rs [hunk 1]
```diff
@@ -130,32 +130,33 @@ pub fn halve_u64<const P: u64>(input: u64) -> u64 {
...
-    for val in vals.iter().rev() {
-        res = res * alpha + TF::from_canonical_u64(val.as_canonical_u64());
+/// Given a slice of SF elements, reduce them to a TF element using a 2^32-base decomposition.
+pub fn reduce_32<SF: PrimeField32, TF: PrimeField>(vals: &[SF]) -> TF {
+    let po2 = TF::from_canonical_usize(1usize << 32);
+    let mut power = TF::one();
+    let mut result = TF::zero();
+    for val in vals {
+        result += TF::from_canonical_u32(val.as_canonical_u32()) * power;
+        power *= po2;
```
Comment (dlubarov) [inline field/src/helpers.rs:140]: Not sure if performance really matters here, but the previous Horner-style method involving `rev()` seems a bit more nicer for a single mul per iteration, is there a downside to that?

File: field/src/helpers.rs [hunk 2]
```diff
@@ -130,32 +130,33 @@ pub fn halve_u64<const P: u64>(input: u64) -> u64 {
...
-/// Given a slice of SF elements, reduce them to a TF element.
-pub fn reduce_64<SF: PrimeField64, TF: PrimeField>(vals: &[SF]) -> TF {
-    let alpha = TF::from_canonical_u64(SF::ORDER_U64);
-
-    let mut res = TF::zero();
-    for val in vals.iter().rev() {
-        res = res * alpha + TF::from_canonical_u64(val.as_canonical_u64());
+/// Given a slice of SF elements, reduce them to a TF element using a 2^32-base decomposition.
+pub fn reduce_32<SF: PrimeField32, TF: PrimeField>(vals: &[SF]) -> TF {
+    let po2 = TF::from_canonical_usize(1usize << 32);
```
Comment (dlubarov) [inline field/src/helpers.rs:135]: Could overflow on 32-bit systems, maybe use `u64` instead of `usize`?

File: field/src/helpers.rs [hunk 3]
```diff
@@ -130,32 +130,33 @@ pub fn halve_u64<const P: u64>(input: u64) -> u64 {
...
-
-        // Can assume there is one u64 digit since SF is PrimeField64.
-        res.push(TF::from_canonical_u64(rem.to_u64_digits()[0]));
+    let mut result = Vec::new();
+    for _ in 0..n {
+        let mask: BigUint = po2.clone() - BigUint::from(1usize);
+        let digit: BigUint = val.clone() & mask;
+        let digit_u32s = digit.to_u32_digits();
+        if !digit_u32s.is_empty() {
+            result.push(TF::from_wrapped_u32(digit.to_u32_digits()[0]));
```
Comment (dlubarov) [inline field/src/helpers.rs:155]: second call to `to_u32_digits` seems redundant?

File: field/src/helpers.rs [hunk 4]
```diff
@@ -130,32 +130,33 @@ pub fn halve_u64<const P: u64>(input: u64) -> u64 {
...
+        } else {
+            result.push(TF::zero())
+        }
+        val /= po2.clone();
```
Comment (dlubarov) [inline field/src/helpers.rs:159]: nit: could use `div_2exp_u64`

**Comment by dlubarov** [comment]
Left a few minor comments, but my one real concern is that `bits % P` where `bits` is a random `u32` seems to be a biased source of randomness. Like most elements of BabyBear will have two `bits` mapped to them, but some will have three.

This should only slightly affect soundness of polynomial identity testing, and I'm not sure ZK will be used here (if gnark will add ZK anyway). But if we go this route, maybe there should be some kind of warning that the user should think through any bias complications.

**Comment by jtguibas** [comment]
@dlubarov That's a good catch. I agree with you that it's biased. I added a SAFETY note to `MultiField32Challenger` about the bias complications for now.

This would be worse in R1CS arithmetizations because additions are free, but one way we could fix this is to instead use the generator `g` of F and compute `g^bits` in the challenger instead. I don't think would actually be much more expensive and if we decide to use PLONK the cost would be equivalent at least for Gnark (they only use one mul add gate).

**Comment by jtguibas** [comment]
One thing is that this soundness issue seems to be part of a larger class of problems where you just want to go from bits to a uniformly random field element. I'm curious if there are circuit-friendly techniques we could use to alleviate the issue.

**Comment by jtguibas** [comment]
It seems like one easy way we could fix this is to sample 32+N bits, where N is small. This will "average out" the number of elements which map to each BabyBear element.

**Comment by dlubarov** [comment]
Yeah that would work to mitigate bias. Maybe we don't need to though... I guess this shouldn't affect ZK, since blinding factors would just come from the prover itself, not from a Challenger. I think the only current uses of challenges are
- Polynomial identity testing, like sampling `z` to test `constraints(z) = quotient(z) * Z_H(z)`. It seems like if our worst-case bias is some elements being sampled with twice the probability of others, a PIT would lose less than one bit of soundness, which seems okay.
- Folding purported codewords in FRI - it's not quite a PIT, but I think it has the same worst case of one bit of soundness loss. We still have a small bounded set of possible challenges which would result in a "bad event" (as the FRI paper put it), and bias would basically enlarge that size by less than one bit.
- FRI query indices - I assume `sample_bits` won't need to go through the field conversion code so there won't have any bias here? If there was bias here though, that would be more of an issue, since the chance of each query missing improperly folded elements can be as high as 1/2 (hence we need extensive parallel repetition to boost this poor soundness); if 1/2 can change to something like 2/3 due to bias we'd really need to adjust the query count.

**Comment by jtguibas** [comment]
> FRI query indices - I assume sample_bits won't need to go through the field conversion code so there won't have any bias here?

Thanks for the detailed writeup. In the current implementation it does actually go through the field conversion code, although this can be easily changed. It also comes with the con of requiring around 4x more permutation calls in the challenger.

I'm also pretty open to also just always reducing a PF into a F to mitigate the bias issue. I don't have concrete numbers on this, but I think the number of permutation calls the challenger makes is quite insignificant compared to FRI itself?

I'll explore both approaches and update the PR.

**Comment by jtguibas** [comment]
Hi @dlubarov, I decided to instead reduce samples of 2^64 into elements of PrimeField32

I haven't done a formal analysis here, but it seems like then the bias will be roughly at most 1/2^32 which should be negligible given the field we're operating over is of size roughly 2^32.

**Comment by dlubarov** [comment]
Yeah that should work, SGTM

**Pull request merged**